avocado/utils/asset.py

```diff
@@ -454,27 +456,50 @@ def find_asset_file(self, create_metadata=False):
         :raises: OSError
         """
 
-        for cache_dir in self.cache_dirs:
-            cache_dir = os.path.expanduser(cache_dir)
-            asset_file = os.path.join(cache_dir, self.relative_dir)
-
-            # Ignore non-files
-            if not os.path.isfile(asset_file):
-                continue
-
-            # Ignore expired asset files
-            if self._is_expired(asset_file, self.expire):
-                continue
-
-            # Ignore mismatch hash
-            if not self._has_valid_hash(asset_file, self.asset_hash, self.algorithm):
-                continue
+        # Try 1: Check if self.name is already a direct file path
+        if self.name and os.path.isfile(self.name):
+            # For direct paths, expired/invalid files are errors, not skips
+            if self._is_expired(self.name, self.expire):
+                raise OSError(f"File {self.asset_name} has expired.")
+            if not self._has_valid_hash(self.name, self.asset_hash, self.algorithm):
+                raise OSError(f"File {self.asset_name} hash mismatch.")
 
             if create_metadata:
-                self._create_metadata_file(asset_file)
-
-            LOG.info("Asset already exists in cache.")
-            return asset_file
+                self._create_metadata_file(self.name)
+            LOG.info("Asset found as direct file path.")
+            return self.name
+
+        # Try 2 & 3: Search in cache directories
+        search_paths = [
+            # Try relative_dir approach first
+            *[
+                (
+                    os.path.join(os.path.expanduser(cache_dir), self.relative_dir),
+                    "Asset already exists in cache.",
+                )
+                for cache_dir in self.cache_dirs
+            ],
+            # Then try by name
```

Comment:
The `self.relative_dir` is already based on `self.name` why it is needed to add search by `self.name`?